### ISSUE TITLE
rust compiler completions syntax/grammar error

### ISSUE THREAD

**Issue by mqudsi** [fish-shell/fish-shell#5057]
While typing in `rustc ./foo`, and error is emitted as the completions for rust are evaluated:

```fish
mqudsi@Blitzkrieg /m/d/r/fish> rustc ./complete: Completion '`flowgraph=<nodeid>`' contained a syntax error
complete: Expected an argument, but found a redirection
`flowgraph=<nodeid>`
```

(caused by `eval`)

**Comment by faho** [comment]
>(caused by eval)

Where are our rustc completions running eval?

**Comment by mqudsi** [comment]
Isn’t the entirety of the argument parameter `eval`’d?

**Comment by faho** [comment]
Ah, that's what you mean! Yes, it's evaluated. I wouldn't say `eval` here, because it's not literally running the `eval`  function.

So, I'd imagine you have a different `rustc` version than I have (nightly?), which explains why I can't reproduce this with the actual rustc completions.

I can however do

```fish
complete -c foo -a 'flowgraph=<nodeid>'
```

and get the error.

Our rustc completions do a whole bunch of

```fish
set flag (string replace -r '^([a-z\-]+\=|[a-z\-]+)(.*)' '$1' $docs[1])
complete -c rustc -x -s C -l codegen -a "$flag" -d "$docs[2]"
```

There's two things we could do here:

- `string escape` the flag

- Loosen the `replace` regex so we remove the `<nodeid>` bit as well

We _need_ to do the latter because a literal `flowgraph=<nodeid>` isn't a usable option (I'm assuming). The question is whether we should also do the former.

(Also, it's weird to me that this is now using `<nodeid>`, when my rustc just uses `metadata=val`)

**Comment by mqudsi** [comment]
Yup, that’s what I was referring to though I didn’t realize it wasn’t ultimately `eval`’d.  I didn’t dig into it too deeply tbh, but yes I’m running rust nightly. The error is one of those that happens only the first time the completion script is sourced and doesn’t show up again in the same fish session. 

I did try to add a `| string escape` for the `$flag` assignment in both the loops (normal and nightly) but I was surprised when that didn’t work and didn’t get a chance to explore it further. (I kept the `“$flag”` substitution quoted even after escapes, too)

I’m surprised this isn’t something we run into more often.
